xds/internal/xdsclient/creds/tls.go

```diff
@@ -0,0 +1,179 @@
+	//
+	// > The only difference between the file-watcher certificate provider
+	// > config and this one is that in the file-watcher certificate provider,
+	// > at least one of the "certificate_file" or "ca_certificate_file" fields
+	// > must be specified, whereas in this configuration, it is acceptable to
+	// > specify neither one.
+	//
+	// We only use a file_watcher provider if either one of them or both are
+	// specified.
+	if cfg.CACertificateFile != "" || cfg.CertificateFile != "" || cfg.PrivateKeyFile != "" {
+		// file_watcher currently ignores BuildOptions.
+		provider, err := certprovider.GetProvider("file_watcher", jd, certprovider.BuildOptions{})
+		if err != nil {
+			// GetProvider fails if jd is invalid, e.g. if only one of private
+			// key and certificate is specified.
+			return nil, fmt.Errorf("failed to get TLS provider: %w", err)
+		}
+		if cfg.CertificateFile != "" {
+			tlsConfig.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+				// Client cert reloading for mTLS.
+				km, err := provider.KeyMaterial(context.Background())
+				if err != nil {
+					return nil, err
+				}
+				if len(km.Certs) != 1 {
+					// xDS bootstrap has a single private key file, so there
+					// must be exactly one certificate or certificate chain
+					// matching this private key.
+					return nil, fmt.Errorf("certificate_file must contains exactly one certificate or certificate chain")
+				}
+				return &km.Certs[0], nil
+			}
+			if cfg.CACertificateFile == "" {
+				// No need for a callback to load the CA. Use the normal mTLS
+				// transport credentials.
+				return &tlsBundle{
+					jd:                   jd,
+					transportCredentials: credentials.NewTLS(&tlsConfig),
+				}, nil
+			}
+		}
+		return &tlsBundle{
+			jd: jd,
+			transportCredentials: &caReloadingClientTLSCreds{
+				baseConfig: &tlsConfig,
+				provider:   provider,
+			},
+		}, nil
+	}
+
+	// None of certificate_file and ca_certificate_file are set.
+	// Use the system-wide root certs.
+	return &tlsBundle{
+		jd:                   jd,
+		transportCredentials: credentials.NewTLS(&tlsConfig),
+	}, nil
+}
+
+func (t *tlsBundle) TransportCredentials() credentials.TransportCredentials {
+	return t.transportCredentials
+}
+
+func (t *tlsBundle) PerRPCCredentials() credentials.PerRPCCredentials {
+	// mTLS provides transport credentials only. There are no per-RPC
+	// credentials.
+	return nil
+}
+
+func (t *tlsBundle) NewWithMode(_ string) (credentials.Bundle, error) {
+	return NewTLS(t.jd)
+}
+
+// caReloadingClientTLSCreds is credentials.TransportCredentials for client
+// side mTLS that attempts to reload the server root certificate from its
+// provider on every client handshake. This is needed because Go's tls.Config
+// does not support reloading the root CAs.
+type caReloadingClientTLSCreds struct {
+	mu         sync.Mutex
+	provider   certprovider.Provider
+	baseConfig *tls.Config
```

Comment:
Would the following work instead?

- `tlsBundle` would have two fields:
  - `identityProvider`: for providing identity key material
  - `rootProvider`: for providing root key material

- When parsing the configuration in `NewTLS`, we will have the following conditions:
  - identity cert/key file name is missing, root file name is missing
    - return `&tlsBundle{}`
  - identity cert/key file name is missing, root file name is present
    - `rp := certprovider.GetProvider(...) // create a pemfile provider for the given root file name`
    - return `&tlsBundle{rootProvider: rp}`
  - identity cert/key file name is present, root file name is missing
    - `ip := certprovider.GetProvider(...) // create a pemfile provider for the given identity cert/key file name`
    - return `&tlsBundle{identityProvider: ip}`
  - identity cert/key file name is present, root file name is present
    - `rp := certprovider.GetProvider(...) // create a pemfile provider for the given root file name`
    - `ip := certprovider.GetProvider(...) // create a pemfile provider for the given identity cert/key file name`
    - return `&tlsBundle{rootProvider: rp, identityProvider: ip}`

```golang
func (t *tlsBundle) ClientHandshake(ctx context.Context, authority string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
	// Create a new `tls.Config` by calling `KeyMaterial` of non-nil root and identity providers
  	
	// Perform the TLS handshake with the tls.Config that we have. We run the
	// actual Handshake() function in a goroutine because we need to respect the
	// deadline specified on the passed in context, and we need a way to cancel
	// the handshake if the context is cancelled.
	conn := tls.Client(rawConn, cfg)
	errCh := make(chan error, 1)
	go func() {
		errCh <- conn.Handshake()
		close(errCh)
	}()
	select {
	case err := <-errCh:
		if err != nil {
			conn.Close()
			return nil, nil, err
		}
	case <-ctx.Done():
		conn.Close()
		return nil, nil, ctx.Err()
	}
	info := credentials.TLSInfo{
		State: conn.ConnectionState(),
		CommonAuthInfo: credentials.CommonAuthInfo{
			SecurityLevel: credentials.PrivacyAndIntegrity,
		},
		SPIFFEID: credinternal.SPIFFEIDFromState(conn.ConnectionState()),
	}
	return credinternal.WrapSyscallConn(rawConn, conn), info, nil
}
```

The above code snippet is copied from the xds credentials implementation which also uses certificate providers.

With this approach we will not have any mutexes in the `tlsBundle` and we won't need a separate `caReloadingClientTLSCreds` type.